pkg/scheduler/ugm/user_tracker_test.go

```diff
@@ -302,6 +302,34 @@ func TestSetAndClearMaxLimits(t *testing.T) {
 	assert.Equal(t, si.EventRecord_UG_USER_LIMIT, eventSystem.Events[1].EventChangeDetail)
 }
 
+func TestUTCanRunApp(t *testing.T) {
+	manager := GetUserManager()
+	defer manager.ClearConfigLimits()
+	user := security.UserGroup{User: "test", Groups: []string{"test"}}
+	userTracker := newUserTracker(user.User, newUGMEvents(mock.NewEventSystemDisabled()))
+	maxRes := resources.NewResourceFromMap(map[string]resources.Quantity{
+		"cpu": 1000,
+	})
+	manager.userWildCardLimitsConfig["root.parent"] = &LimitConfig{
+		maxResources:    maxRes,
+		maxApplications: 3,
+	} // manipulate map directly to avoid hierarchy creation
+
+	hierarchy1 := strings.Split(path1, configs.DOT)
+	assert.Assert(t, userTracker.canRunApp(hierarchy1, TestApp1))
+	// make sure wildcard limits are applied
+	assert.Equal(t, uint64(3), userTracker.queueTracker.childQueueTrackers["parent"].maxRunningApps)
+	assert.Assert(t, resources.Equals(maxRes, userTracker.queueTracker.childQueueTrackers["parent"].maxResources))
+
+	// limit hit
+	userTracker.setLimits(path1, resources.Zero, 1, false, false)
+	userTracker.increaseTrackedResource(path1, TestApp1, resources.NewResourceFromMap(map[string]resources.Quantity{
+		"cpu": 1000,
```

Comment:
The cpu resource is not a part of limit in this test case, right? If so, I'm a bit confused why `canRunApp` does not check max resource?

Reply:
The limit is maxApps = 1 set by `setLimits()`. You "register" the running application with `increaseTrackedResource()`, there's no separate method for that. Although using `resources.Zero` is a bit misleading as a maximum resource, I'll fix that (it just happens to work right now because we don't check this during increment and I even created a JIRA to remove the boolean in YUNIKORN-2423, we can add this increment verification later).